### PR TITLE
Support for models having partially non trainable parameters

### PR DESCRIPTION
**Pull request by baijumeswani**
**Description**:
This pull request brings support for models where only a part of the parameters are trainable. A simple example of such a model is:

```py
class NeuralNetNoGradModel(torch.nn.Module):
    def __init__(self, input_size, hidden_size, num_classes):
        super(NeuralNetNoGradModel, self).__init__()

        self.fc1 = torch.nn.Linear(input_size, hidden_size).requires_grad_(False)
        self.relu = torch.nn.ReLU()
        self.fc2 = torch.nn.Linear(hidden_size, num_classes)

    def forward(self, model_input):
        out = self.relu(self.fc1(model_input))
        out = self.fc2(out)
        return out
```

In the example above, the ```fc1``` parameter will not be trained.

One of the known limitation of ```ORTModule``` is that there is currently no support for models where a part of the model is executed under the ```torch.no_grad()``` context. For example, the below model does not have support:

```py
class NeuralNetNoGradModel(torch.nn.Module):
    def __init__(self, input_size, hidden_size, num_classes):
        super(NeuralNetNoGradModel, self).__init__()

        self.fc1 = torch.nn.Linear(input_size, hidden_size).requires_grad_(False)
        self.relu = torch.nn.ReLU()
        self.fc2 = torch.nn.Linear(hidden_size, num_classes)

    def forward(self, model_input):
        out = None
        with torch.no_grad():
            out = self.relu(self.fc1(model_input))
        out = self.fc2(out)
        return out
```

With this example, ```ORTModule``` is not able to determine which parameters are trainable and which are not. As a work around, users would need to define the model as shown in the very first example (by setting the ```requires_grad``` attribute of parameters.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule.py [hunk 1]
```diff
@@ -241,9 +243,17 @@ def backward(ctx, *grad_outputs):
...
                             # Append None to results for each input that did not require grad
                             results.append(None)
+
                     # Append gradients of initializer to results
-                    results += [_ort_output_to_torch_tensor(backward_output)
-                                for backward_output in backward_outputs[num_user_input_grads:]]
+                    # Go over each initializer, check if it required grad and append to results accordingly
+                    initializer_index = 0
+                    for inititializer_name in self._onnx_graphs_info.initializer_names:
+                        if inititializer_name in self._onnx_graphs_info.initializer_names_to_train:
```
Comment (mrry) [inline orttraining/orttraining/python/training/ortmodule.py:251]: Is this loop O(n^2) in the number of initializers, since `initializer_name in self._onnx_graphs_info.initializer_names_to_train` does linear search in the list? 

Would be good to figure out if we can reduce the complexity and/or cache the mapping work at the same time as export?

File: orttraining/orttraining/python/training/ortmodule.py [hunk 2]
```diff
@@ -241,9 +243,17 @@ def backward(ctx, *grad_outputs):
                             # input_name is not found in the self._input_names_require_grad list
                             # Append None to results for each input that did not require grad
                             results.append(None)
+
                     # Append gradients of initializer to results
-                    results += [_ort_output_to_torch_tensor(backward_output)
-                                for backward_output in backward_outputs[num_user_input_grads:]]
+                    # Go over each initializer, check if it required grad and append to results accordingly
+                    initializer_index = 0
+                    for inititializer_name in self._onnx_graphs_info.initializer_names:
```
Comment (mrry) [inline orttraining/orttraining/python/training/ortmodule.py:250]: Typo: s/inititalizer_name/initializer_name/

File: orttraining/orttraining/python/training/ortmodule.py [hunk 3]
```diff
@@ -332,13 +342,18 @@ def _initialize_module_gradient_graph_builder(self):
...
             p.name for p in self._onnx_inference.graph.initializer}
-        initializer_names = [
-            p for p in initializer_names if p in onnx_initializer_names]
+        initializer_names_to_train = [
+            p for p in initializer_names_to_train if p in onnx_initializer_names]
 
         # Build full training graph
         grad_builder_config = C.ModuleGradientGraphBuilderConfiguration()
+        grad_builder_config.initializer_names = initializer_names
         grad_builder_config.initializer_names_to_train = initializer_names
```
Comment (mrry) [inline orttraining/orttraining/python/training/ortmodule.py:357]: Should this be set to `initializer_names_to_train`? (Commenting on the before copy, since the line wasn't part of the diff and so GitHub won't let me comment on this line in the new version :).)

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 4]
```diff
@@ -1311,3 +1311,41 @@ def forward(self, input1):
 
     assert output['out'] is not None
     assert output['none_output'] is None
+
+def test_model_partially_requires_grad():
```
Comment (mrry) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1315]: Can you add a test where the `requires_grad_` is called on one of the submodules after the first call to `ORTModule.forward()`?

**Review by mrry — CHANGES_REQUESTED** [review]

File: orttraining/orttraining/python/training/ortmodule.py (hunk 3 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule.py:357]: yes, this was the result of the merging the dev on master and me resolving my conflicts. Will fix. Thanks.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 4 above)
Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1315]: Do you mean when ```requires_grad_``` chacges between one ORTModule forward call to another?

Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1328]: I've added support for the scenario where users may want to change the requires_grad attribute in the middle of training. And have also added a unit test for the same.
Let me know if this is not what you intended.

File: orttraining/orttraining/python/training/ortmodule.py [hunk 5]
```diff
@@ -107,7 +107,9 @@ def _forward(self, *inputs, **kwargs):
             Finally, we instantiate the ONNX Runtime InferenceSession.
             '''
             # TODO: using pytorch for evaluation for now. We will use ORT for evaluation later.
-            if not self._is_training:
+            # TODO: If the model is being executed with the gradient disabled (inside torch.no_grad() context for example),
+            # leverage pytorch model for now.
+            if not self._is_training or not torch.is_grad_enabled():
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:112]: Instead of having 2 places to indicate the model is in training mode, we can upgrade `self._is_training` to a `@property` and in the getter implementation we add the `torch.is_grad_enabled()`, having a single point of check

File: orttraining/orttraining/python/training/ortmodule.py [hunk 6]
```diff
@@ -232,9 +242,17 @@ def backward(ctx, *grad_outputs):
...
-                    results += [_ortvalue_to_torch_tensor(backward_output)
-                                for backward_output in backward_outputs[num_user_input_grads:]]
+                    # Go over each initializer, check if it required grad and append to results accordingly
+                    initializer_index = 0
+                    for inititializer_name in self._onnx_graphs_info.initializer_names:
+                        if inititializer_name in self._onnx_graphs_info.initializer_names_to_train:
+                            results.append(_ortvalue_to_torch_tensor(backward_outputs[num_user_input_grads+initializer_index]))
+                            initializer_index += 1
+                        else:
+                            results.append(None)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:254]:
```suggestion
                    initializer_index = num_user_input_grads
                    for inititializer_name in self._onnx_graphs_info.initializer_names:
                        if inititializer_name in self._onnx_graphs_info.initializer_names_to_train:
                            results.append(_ortvalue_to_torch_tensor(backward_outputs[initializer_index]))
                            initializer_index += 1
                        else:
                            results.append(None)
```

File: orttraining/orttraining/python/training/ortmodule.py [hunk 7]
```diff
@@ -122,16 +124,24 @@ def _forward(self, *inputs, **kwargs):
...
                     self._original_module_parameters, self._onnx_inference, *inputs, **kwargs)
+            initializer_names_to_train_set = {initializer[0] for initializer in
+                self._flattened_output_module.named_parameters() if initializer[1].requires_grad}
             # If inputs requiring gradient change from forward to the next, the module_gradient_graph_builder
             # needs to be reinitialized so it can compute the backward output for the new inputs that require_grad
-            if input_names_require_grad != self._input_names_require_grad:
+            if input_names_require_grad != self._input_names_require_grad or \
+                initializer_names_to_train_set != self._initializer_names_to_train_set:
                 self._input_names_require_grad = input_names_require_grad
+                self._initializer_names_to_train_set = initializer_names_to_train_set
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:139]: Any way we can avoid having `self._initializer_names_to_train` and `self._initializer_names_to_train_set`?

Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule.py:139]: ```self._initializer_names_to_train``` is not a direct property. It is one that is accessed through the onnx graph populated by the backend through ```self._onnx_graphs_info.initializer_names_to_train```.
In order to keep the lookup time linear, I need to have a set to look up from. But the order is also important for building the backward results which is why, we need both.

File: orttraining/orttraining/python/training/ortmodule.py (hunk 5 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule.py:112]: yeah that is a good idea. I'll do that.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 8]
```diff
@@ -1359,3 +1372,49 @@ def forward(self, mask, x1, x2):
...
+    # Make sure no exception is raised
+    output = model(x)
+    loss = torch.sum(output)
+    loss.backward()
+    training_session1 = model._training_session
+
+    model._original_module.fc1.requires_grad_(False)
+    output = model(x)
+    loss = torch.sum(output)
+    loss.backward()
```
Comment (mrry) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1417]: Is there some way we could test that this is having the intended effect? e.g. Before the first call, assert that `model.fc1.grad` is zero, then non-zero after the first call, and the same non-zero value after the second call? (I'm assuming that `requires_grad_(False)` doesn't release the accumulated grad buffer....)

**Review by mrry — COMMENTED** [review]
Thanks for the update! Just a few small nits/suggestions.

File: orttraining/orttraining/python/training/ortmodule.py [hunk 9]
```diff
@@ -122,16 +124,24 @@ def _forward(self, *inputs, **kwargs):
                 self._get_inference_graph_and_init_gradient_graph_builder(
                     *inputs, **kwargs)
 
+            # Flag to indicate whether the gradient_graph needs to be built
+            build_gradient_graph = self._current_input_shape is None
             _, _, input_names_require_grad, new_input_shape = \
                 _ortmodule_io.parse_inputs_for_onnx_export(
                     self._original_module_parameters, self._onnx_inference, *inputs, **kwargs)
+            initializer_names_to_train_set = {initializer[0] for initializer in
+                self._flattened_output_module.named_parameters() if initializer[1].requires_grad}
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule.py]: addressed this.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 8 above)
Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1466]: I added the asserts to test that.

File: orttraining/orttraining/python/training/ortmodule.py [hunk 10]
```diff
@@ -325,18 +347,31 @@ def backward(ctx, *grad_outputs):
...
+        self._is_training = training_mode
+
     def _initialize_module_gradient_graph_builder(self):
         # TODO: PyTorch exporter bug: changes the initializer order in ONNX model
-        initializer_names = [p[0]
-                             for p in self._flattened_output_module.named_parameters()]
+        initializer_names = [name
+                             for name, _ in self._flattened_output_module.named_parameters()]
+        initializer_names_to_train = []
+        if torch.is_grad_enabled():
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:363]: should it be `self.is_training`?

File: orttraining/orttraining/python/training/ortmodule.py [hunk 11]
```diff
@@ -287,7 +308,7 @@ def backward(ctx, *grad_outputs):
                     "The model's forward method has **kwargs parameter which is currently not supported.")
 
         self._onnx_inference = None
-        self._is_training = True
+        self.is_training = True
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:311]: self._is_training is not initialized at __init__?

Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule.py:311]: it is being initialized by the setter. It would defeat the purpose of setting it here since we want it to be only set in the setter method.

File: orttraining/orttraining/python/training/ortmodule.py (hunk 10 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule.py:363]: yeah, I think this should be is_training. Updating.

File: orttraining/orttraining/python/training/ortmodule.py [hunk 12]
```diff
@@ -122,16 +124,25 @@ def _forward(self, *inputs, **kwargs):
...
                     *inputs, **kwargs)
 
+            # Flag to indicate whether the gradient_graph needs to be built
+            build_gradient_graph = self._current_input_shape is None
             _, _, input_names_require_grad, new_input_shape = \
                 _ortmodule_io.parse_inputs_for_onnx_export(
                     self._original_module_parameters, self._onnx_inference, *inputs, **kwargs)
+            initializer_names_to_train_set = {name for name, param in
+                self._flattened_output_module.named_parameters() if param.requires_grad}
+            initializer_names_to_train_set_prev = set(self._onnx_graphs_info.initializer_names_to_train) \
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:134]: Regarding to the naming, initializer_names_to_train_set_prev actually refers to the current list of initializers (not prev) and  initializer_names_to_train_set refers to the new initializers?

I mean, using the onnx graph as ground truth, any property at self._onnx_graphs_info is the current value. Anything coming through forward is a possibly new thing

File: orttraining/orttraining/python/training/ortmodule.py (hunk 11 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule.py:311]: Current implementatio sets it as True at https://github.com/microsoft/onnxruntime/blob/master/orttraining/orttraining/python/training/ortmodule.py#L290

A model is in training model by default. we dont have to call model.train() to put it training, right?

File: orttraining/orttraining/python/training/ortmodule.py (hunk 12 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule.py]: renamed it as ```initializer_names_to_train_set_user_model``` and ```initializer_names_to_train_set_onnx_graph``` to avoid any confusion.

**Review by thiagocrepaldi — DISMISSED** [review]

File: orttraining/orttraining/python/training/ortmodule.py (hunk 11 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule.py:311]: I think this was set with the assumption that user model is in training mode. What we can do is, set the training mode to the training mode of the original user model. Or better still, not have a flag of our own. We just leverage the original_model training flag.
I can make these changes.

**Review by mrry — APPROVED** [review]

**Review by thiagocrepaldi — APPROVED** [review]

**Comment by baijumeswani** [comment]
Thank you for the review. :)
